src/main/java/com/zaxxer/hikari/HikariConfig.java

```diff
@@ -1166,10 +1168,13 @@ private String generatePoolName()
       final var prefix = "HikariPool-";
       try {
          // Pool number is global to the VM to avoid overlapping pool numbers in classloader scoped environments
-         synchronized (System.getProperties()) {
+         HIKARI_CONFIG_LOCK.lock();
```

Comment:
There is a ticket (#2104) that describes a bug with the existing implementation. I would like a solution that solves that bug that also satisfies the virtual thread requirements. I had a terrible idea for a solution that would be something like this:

```java
   private String generatePoolName()
   {
      final var prefix = "HikariPool-";

      String failureReason;
      // Pool number is global to the VM to avoid overlapping pool numbers in classloader scoped environments
      try {
         // create a unique ID to identify ourselves for the locking semantics below
         String lockId = UUID.randomUUID().toString();

         // get the start time so that we can check how long we have been running and implement a timeout.
         long startTime = System.currentTimeMillis();

         // This code is essentially reimplementing lock semantics using system properties in order to get an atomically
         // incrementing pool number. We can't use a global lock since that won't work across classloaders.
         do {
            String lock = System.getProperty("com.zaxxer.hikari.pool_number.lock");
            // if the lock is null, then another thread is not definitely running.
            if (lock == null) {
               // set the lock to our ID to indicate that we want to start work
               System.setProperty("com.zaxxer.hikari.pool_number.lock", lockId);

               // get the lock again to check that no other thread has
               lock = System.getProperty("com.zaxxer.hikari.pool_number.lock");
               if (lockId.equals(lock)) {
                  try {
                     final var next = String.valueOf(Integer.getInteger("com.zaxxer.hikari.pool_number", 0) + 1);
                     System.setProperty("com.zaxxer.hikari.pool_number", next);
                     return prefix + next;
                  }
                  finally {
                     System.clearProperty("com.zaxxer.hikari.pool_number.lock");
                  }
               }
            }

            // If we are here it means there is another thread trying to set the pool number. Let's tell the
            // JVM to let other threads run first since we are waiting and we don't want to spin too fast.
            Thread.yield();
         } while (System.currentTimeMillis() - startTime < 5000L);

         failureReason = "timed out waiting for lock on pool numbering";
      } catch (AccessControlException e) {
         failureReason = "security manager prevented access to system properties";
      }

      // The SecurityManager didn't allow us to read/write system properties
      // so just generate a random pool number instead
      final var random = ThreadLocalRandom.current();
      final var buf = new StringBuilder(prefix);

      for (var i = 0; i < 4; i++) {
         buf.append(ID_CHARACTERS[random.nextInt(62)]);
      }

      LOGGER.info("assigned random pool name '{}' ({})", buf, failureReason);

      return buf.toString();
   }
   ```

Reply:
@lfbayer you are right and I reverted this file back to use the synchronized block

Reply:
But this code is also bad, as I believe I already see a race condition.